cli/lsp/completions.rs

```diff
@@ -452,6 +463,114 @@ fn get_relative_specifiers(
     .collect()
 }
 
+/// Get completions for `npm:` specifiers.
+async fn get_npm_completions(
+  specifier: &str,
+  range: &lsp::Range,
+  npm_registry_api: &impl NpmRegistryApi,
+  npm_search_api: &impl NpmSearchApi,
+) -> Option<Vec<lsp::CompletionItem>> {
+  debug_assert!(specifier.starts_with("npm:"));
+  let bare_specifier = &specifier[4..];
+
+  // Find the index of the '@' delimiting the package name and version, if any.
+  let v_index = if bare_specifier.starts_with('@') {
+    bare_specifier
+      .find('/')
+      .filter(|idx| !bare_specifier[1..*idx].is_empty())
+      .and_then(|idx| {
+        bare_specifier[idx..]
+          .find('@')
+          .filter(|idx2| !bare_specifier[(idx + 1)..*idx2].is_empty())
+          .filter(|idx2| !bare_specifier[(idx + 1)..*idx2].contains('/'))
+      })
+  } else {
+    bare_specifier
+      .find('@')
+      .filter(|idx| !bare_specifier[..*idx].is_empty())
+      .filter(|idx| !bare_specifier[..*idx].contains('/'))
+  };
+
+  // First try to match `npm:some-package@<version-to-complete>`.
+  if let Some(v_index) = v_index {
+    let package_name = &bare_specifier[..v_index];
+    let v_prefix = &bare_specifier[(v_index + 1)..];
+    let versions = &npm_registry_api
+      .package_info(package_name)
```

Comment:
The implementation of this currently uses the cached package information in the LSP. It would be good to make the `NpmSearchApi` trait have a `async fn package_versions(&self) -> Option<Vec<Version>>` method so that this is not coupled to `deno_npm`'s trait and then we could implement a more targeted search api that also uses the module registry's file fetcher.